### DISCUSSION TITLE
Identifying memory leaks in nextflow workflow

### DISCUSSION THREAD

**Discussion by anoronh4** [nextflow-io/nextflow#1944] [General]
We are looking to make the main nextflow run more memory efficient by identifying memory leaks and objects that require a lot of the heap. However we are unsure of where to start and have not found a lot of guidance on how to best address it specifically for nextflow. We are processing >10,000 and at one process, a sample will spawn about 60 jobs for variant calling different parts of the genome. this means our runs are submitting a lot of jobs and we also have a large amount of channels that don't close because we are using `watchPath` on up to 4 files at a time. the maps are few, quite small and updated infrequently. 

We are already fine-tuning `NXF_OPTS` and heap size variables, but this seems to just be postponing the problem to another day instead of addressing underlying issues in the workflow. any guidance would be appreciated. 

finally i do want to mention that i have a TimerTask in my workflow:
```
def touchInputs() {
  new Timer().schedule({
  for ( i in epochMap.keySet() ){
    fileEpoch = file(i).lastModified()
    if ( fileEpoch > epochMap[i]) {
      epochMap[i] = fileEpoch
      "touch -ca ${i}".execute()
    }
  }
} as TimerTask, 1000, params.touchInputsInterval * 60 * 1000 ) // convert minutes to milliseconds
}
```
the map operated on in the above snippet has only four keys in it and i do not believe it to be a memory leak, but maybe someone else would be able to identify something. `touchInputs()` is only run once.

**Comment by pditommaso** [comment]
Too broad context. You can use tools and techniques for Java application to try identifying the source of the leak.
